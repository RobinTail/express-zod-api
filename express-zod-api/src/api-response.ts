import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
