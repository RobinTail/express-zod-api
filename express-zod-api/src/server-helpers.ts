import type fileUpload from "express-fileupload";
import { loadPeer } from "./peer-helpers.ts";
import { AbstractResultHandler } from "./result-handler.ts";
import { ActualLogger } from "./logger-helpers.ts";
import { CommonConfig, ServerConfig } from "./config-type.ts";
import { ErrorRequestHandler, RequestHandler, Request } from "express";
import createHttpError from "http-errors";
import { lastResortHandler } from "./last-resort.ts";
import { ResultHandlerError } from "./errors.ts";
import { ensureError } from "./common-helpers.ts";
import { monitor } from "./graceful-shutdown.ts";

// eslint-disable-next-line no-restricted-syntax -- substituted by TSDOWN
export const localsID = Symbol.for(process.env.TSDOWN_SELF!);

type EquippedRequest = Request<
  unknown,
  unknown,
  unknown,
  unknown,
  { [localsID]?: { logger: ActualLogger } }
>;

/** @desc Returns child logger for the given request (if configured) or the configured logger otherwise */
export type GetLogger = (request?: Request) => ActualLogger;

interface HandlerCreatorParams {
  errorHandler: AbstractResultHandler;
  getLogger: GetLogger;
}

export const createCatcher =
  ({ errorHandler, getLogger }: HandlerCreatorParams): ErrorRequestHandler =>
  async (error, request, response, next) => {
    if (!error) return next();
    return errorHandler.execute({
      error: ensureError(error),
      request,
      response,
      input: null,
      output: null,
      ctx: {},
      logger: getLogger(request),
    });
  };

export const createNotFoundHandler =
  ({ errorHandler, getLogger }: HandlerCreatorParams): RequestHandler =>
  async (request, response) => {
    const error = createHttpError(
      404,
      `Can not ${request.method} ${request.path}`,
    );
    const logger = getLogger(request);
    try {
      await errorHandler.execute({
        request,
        response,
        logger,
        error,
        input: null,
        output: null,
        ctx: {},
      });
    } catch (e) {
      lastResortHandler({
        response,
        logger,
        error: new ResultHandlerError(ensureError(e), error),
      });
    }
  };

export const createUploadFailureHandler =
  (error: Error): RequestHandler =>
  (req, {}, next) => {
    const failedFile = Object.values(req?.files || [])
      .flat()
      .find(({ truncated }) => truncated);
    if (failedFile) return next(error);
    next();
  };

export const createUploadLogger = (
  logger: ActualLogger,
): Pick<Console, "log"> => ({ log: logger.debug.bind(logger) });

export const createUploadParsers = async ({
  getLogger,
  config,
}: {
  getLogger: GetLogger;
  config: ServerConfig;
}): Promise<RequestHandler[]> => {
  const uploader = await loadPeer<typeof fileUpload>("express-fileupload");
  const { limitError, beforeUpload, ...options } = {
    ...(typeof config.upload === "object" && config.upload),
  };
  const parsers: RequestHandler[] = [];
  parsers.push(async (request, response, next) => {
    const logger = getLogger(request);
    await beforeUpload?.({ request, logger });
    return uploader({
      debug: true,
      ...options,
      abortOnLimit: false,
      parseNested: true,
      logger: createUploadLogger(logger),
    })(request, response, next);
  });
  if (limitError) parsers.push(createUploadFailureHandler(limitError));
  return parsers;
};

export const moveRaw: RequestHandler = (req, {}, next) => {
  if (Buffer.isBuffer(req.body)) req.body = { raw: req.body };
  next();
};

/** @since v22.13 the access logging can be customized and disabled */
export const createLoggingMiddleware =
  ({
    logger: parent,
    config: {
      childLoggerProvider,
      accessLogger = ({ method, path }, logger) =>
        logger.debug(`${method}: ${path}`),
    },
  }: {
    logger: ActualLogger;
    config: CommonConfig;
  }): RequestHandler =>
  async (request, response, next) => {
    const logger = (await childLoggerProvider?.({ request, parent })) || parent;
    accessLogger?.(request, logger);
    if (request.res)
      (request as EquippedRequest).res!.locals[localsID] = { logger };
    next();
  };

export const makeGetLogger =
  (fallback: ActualLogger): GetLogger =>
  (request) =>
    (request as EquippedRequest | undefined)?.res?.locals[localsID]?.logger ||
    fallback;

export const installDeprecationListener = (logger: ActualLogger) =>
  process.on("deprecation", ({ message, namespace, name, stack }) =>
    logger.warn(
      `${name} (${namespace}): ${message}`,
      stack.split("\n").slice(1),
    ),
  );

export const installTerminationListener = ({
  servers,
  logger,
  options: { timeout, beforeExit, events = ["SIGINT", "SIGTERM"] },
}: {
  servers: Parameters<typeof monitor>[0];
  options: Extract<ServerConfig["gracefulShutdown"], object>;
  logger: ActualLogger;
}) => {
  const graceful = monitor(servers, { logger, timeout });
  const onTerm = async () => {
    await graceful.shutdown();
    await beforeExit?.();
    process.exit();
  };
  for (const trigger of events) process.on(trigger, onTerm);
};
