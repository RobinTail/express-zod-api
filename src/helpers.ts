import {Request} from 'express';
import {z} from 'zod';
import {MiddlewareDefinition} from './middleware';

export type FlatObject = Record<string, any>;
export type ObjectSchema = z.AnyZodObject;

export type Merge<A extends ObjectSchema, B extends ObjectSchema | any> = z.ZodObject<
  // eslint-disable-next-line @typescript-eslint/ban-types
  A['shape'] & (B extends ObjectSchema ? B['shape'] : {}),
  A['_unknownKeys'],
  A['_catchall']
>;

export function combineEndpointAndMiddlewareInputSchemas<IN extends ObjectSchema, mIN>(
  input: IN,
  middlewares: MiddlewareDefinition<any, any, any>[]
): Merge<IN, mIN> {
  if (middlewares.length === 0) {
    return input as any as Merge<IN, mIN>;
  }
  return middlewares
    .map((middleware) => middleware.input)
    .reduce((carry: ObjectSchema, schema) => carry.merge(schema))
    .merge(input) as Merge<IN, mIN>;
}

export function getInitialInput(request: Request): any {
  switch (request.method) {
    case 'POST':
    case 'PUT':
    case 'PATCH':
      return request.body;
    case 'GET':
      return request.query;
    case 'DELETE': // _may_ have body
    default:
      return {...request.query, ...request.body};
  }
}
