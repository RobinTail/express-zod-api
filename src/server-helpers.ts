import type fileUpload from "express-fileupload";
import { metaSymbol } from "./metadata";
import { loadPeer } from "./peer-helpers";
import { AnyResultHandlerDefinition } from "./result-handler";
import { AbstractLogger } from "./logger";
import { ServerConfig } from "./config-type";
import { ErrorRequestHandler, RequestHandler, Response } from "express";
import createHttpError, { isHttpError } from "http-errors";
import { lastResortHandler } from "./last-resort";
import { ResultHandlerError } from "./errors";
import { makeErrorFromAnything } from "./common-helpers";

interface HandlerCreatorParams {
  errorHandler: AnyResultHandlerDefinition;
  rootLogger: AbstractLogger;
}

export type LocalResponse = Response<
  unknown,
  { [metaSymbol]?: { logger: AbstractLogger } }
>;

export const createParserFailureHandler =
  ({ errorHandler, rootLogger }: HandlerCreatorParams): ErrorRequestHandler =>
  async (error, request, response: LocalResponse, next) => {
    if (!error) {
      return next();
    }
    errorHandler.handler({
      error: isHttpError(error)
        ? error
        : createHttpError(400, makeErrorFromAnything(error).message),
      request,
      response,
      input: null,
      output: null,
      options: {},
      logger: response.locals[metaSymbol]?.logger || rootLogger,
    });
  };

export const createNotFoundHandler =
  ({ errorHandler, rootLogger }: HandlerCreatorParams): RequestHandler =>
  async (request, response: LocalResponse) => {
    const error = createHttpError(
      404,
      `Can not ${request.method} ${request.path}`,
    );
    const logger = response.locals[metaSymbol]?.logger || rootLogger;
    try {
      errorHandler.handler({
        request,
        response,
        logger,
        error,
        input: null,
        output: null,
        options: {},
      });
    } catch (e) {
      lastResortHandler({
        response,
        logger,
        error: new ResultHandlerError(makeErrorFromAnything(e).message, error),
      });
    }
  };

export const createUploadFailueHandler =
  (error: Error): RequestHandler =>
  (req, {}, next) => {
    const failedFile = Object.values(req?.files || [])
      .flat()
      .find(({ truncated }) => truncated);
    if (failedFile) {
      return next(error);
    }
    next();
  };

export const createUploadLogger = (
  logger: AbstractLogger,
): Pick<Console, "log"> => ({
  log: logger.debug.bind(logger),
});

export const createUploadParsers = async ({
  rootLogger,
  config,
}: {
  rootLogger: AbstractLogger;
  config: ServerConfig;
}): Promise<RequestHandler[]> => {
  const uploader = await loadPeer<typeof fileUpload>("express-fileupload");
  const { limitError, beforeUpload, ...options } = {
    ...(typeof config.server.upload === "object" && config.server.upload),
  };
  const parsers: RequestHandler[] = [];
  parsers.push(async (request, response: LocalResponse, next) => {
    const logger = response.locals[metaSymbol]?.logger || rootLogger;
    try {
      await beforeUpload?.({ request, logger });
    } catch (error) {
      return next(error);
    }
    uploader({
      ...options,
      abortOnLimit: false,
      parseNested: true,
      logger: createUploadLogger(logger),
    })(request, response, next);
  });
  if (limitError) {
    parsers.push(createUploadFailueHandler(limitError));
  }
  return parsers;
};

export const moveRaw: RequestHandler = (req, {}, next) => {
  if (Buffer.isBuffer(req.body)) {
    req.body = { raw: req.body };
  }
  next();
};

/** @since v19 prints the actual path of the request, not a configured route */
export const createLoggingMiddleware =
  ({
    rootLogger,
    config,
  }: {
    rootLogger: AbstractLogger;
    config: ServerConfig;
  }): RequestHandler =>
  async (request, response: LocalResponse, next) => {
    const logger = config.childLoggerProvider
      ? await config.childLoggerProvider({ request, parent: rootLogger })
      : rootLogger;
    logger.info(`${request.method}: ${request.path}`);
    response.locals[metaSymbol] = { logger };
    next();
  };
