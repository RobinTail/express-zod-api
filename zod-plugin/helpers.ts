import type { z } from "zod";
import { unpack } from "./packer";

export const getBrand = (subject: z.core.$ZodType) => {
  const { brand } = unpack(subject) || {};
  if (
    typeof brand === "symbol" ||
    typeof brand === "string" ||
    typeof brand === "number"
  )
    return brand;
  return undefined;
};
