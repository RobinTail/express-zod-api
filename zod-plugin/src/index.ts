import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
