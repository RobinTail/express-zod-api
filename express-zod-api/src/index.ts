import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
