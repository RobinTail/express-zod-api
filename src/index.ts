import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
