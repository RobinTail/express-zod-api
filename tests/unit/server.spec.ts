import { moveRaw } from "../../src/server-helpers";
import { givePort } from "../helpers";
import {
  appMock,
  compressionMock,
  expressJsonMock,
  expressMock,
  expressRawMock,
} from "../express-mock";
import {
  createHttpsServerSpy,
  httpListenSpy,
  httpsListenSpy,
} from "../http-mock";
import { z } from "zod";
import {
  AppConfig,
  BuiltinLogger,
  EndpointsFactory,
  ServerConfig,
  attachRouting,
  createServer,
  defaultResultHandler,
  ez,
  createConfig,
} from "../../src";
import express from "express";

describe("Server", () => {
  afterAll(() => {
    vi.restoreAllMocks();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("Express is mocked", () => {
    expect(expressMock).toBeTruthy();
  });

  describe("createServer()", () => {
    test("Should create server with minimal config", async () => {
      const port = givePort();
      const configMock = {
        http: { listen: port },
        cors: true,
        startupLogo: false,
        logger: { level: "warn" as const },
      };
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            methods: ["get", "post"],
            input: z.object({
              n: z.number(),
            }),
            output: z.object({
              b: z.boolean(),
            }),
            handler: vi.fn(),
          }),
        },
      };
      const { servers } = await createServer(
        createConfig(configMock),
        routingMock,
      );
      expect(servers).toHaveLength(1);
      expect(servers[0]).toBeTruthy();
      expect(appMock).toBeTruthy();
      expect(appMock.disable).toHaveBeenCalledWith("x-powered-by");
      expect(appMock.use).toHaveBeenCalledTimes(2);
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/test",
        expressJsonMock,
        expect.any(Function), // endpoint
      );
      expect(appMock.post).toHaveBeenCalledTimes(1);
      expect(appMock.post).toHaveBeenCalledWith(
        "/v1/test",
        expressJsonMock,
        expect.any(Function), // endpoint
      );
      expect(appMock.options).toHaveBeenCalledTimes(1);
      expect(appMock.options).toHaveBeenCalledWith(
        "/v1/test",
        expressJsonMock,
        expect.any(Function), // endpoint
      );
      expect(httpListenSpy).toHaveBeenCalledTimes(1);
      expect(httpListenSpy).toHaveBeenCalledWith(port, expect.any(Function));
    });

    test("Should create server with custom JSON parser, raw parser, logger, error handler and beforeRouting", async () => {
      const customLogger = new BuiltinLogger({ level: "silent" });
      const infoMethod = vi.spyOn(customLogger, "info");
      const port = givePort();
      const configMock = {
        http: { listen: { port } }, // testing Net::ListenOptions
        jsonParser: vi.fn(),
        rawParser: vi.fn(),
        beforeRouting: vi.fn(),
        cors: true,
        startupLogo: false,
        errorHandler: {
          handler: vi.fn(),
        },
        logger: customLogger,
      };
      const factory = new EndpointsFactory(defaultResultHandler);
      const routingMock = {
        v1: {
          test: factory.build({
            methods: ["get", "post"],
            input: z.object({
              n: z.number(),
            }),
            output: z.object({
              b: z.boolean(),
            }),
            handler: vi.fn(),
          }),
          raw: factory.build({
            method: "patch",
            input: ez.raw(),
            output: z.object({}),
            handler: vi.fn(),
          }),
        },
      };
      const { logger, app } = await createServer(
        configMock as unknown as ServerConfig,
        routingMock,
      );
      expect(logger).toEqual(customLogger);
      expect(app).toEqual(appMock);
      expect(appMock).toBeTruthy();
      expect(appMock.use).toHaveBeenCalledTimes(2);
      expect(configMock.errorHandler.handler).toHaveBeenCalledTimes(0);
      expect(configMock.beforeRouting).toHaveBeenCalledWith({
        app: appMock,
        logger: customLogger,
        getChildLogger: expect.any(Function),
      });
      expect(infoMethod).toHaveBeenCalledTimes(1);
      expect(infoMethod).toHaveBeenCalledWith(`Listening`, { port });
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/test",
        configMock.jsonParser,
        expect.any(Function), // endpoint
      );
      expect(appMock.post).toHaveBeenCalledTimes(1);
      expect(appMock.post).toHaveBeenCalledWith(
        "/v1/test",
        configMock.jsonParser,
        expect.any(Function), // endpoint
      );
      expect(appMock.patch).toHaveBeenCalledTimes(1);
      expect(appMock.patch).toHaveBeenCalledWith(
        "/v1/raw",
        configMock.rawParser,
        moveRaw,
        expect.any(Function), // endpoint
      );
      expect(appMock.options).toHaveBeenCalledTimes(2);
      expect(appMock.options).toHaveBeenCalledWith(
        "/v1/test",
        configMock.jsonParser,
        expect.any(Function), // endpoint
      );
      expect(appMock.options).toHaveBeenCalledWith(
        "/v1/raw",
        configMock.rawParser,
        moveRaw,
        expect.any(Function), // endpoint
      );
      expect(httpListenSpy).toHaveBeenCalledTimes(1);
      expect(httpListenSpy).toHaveBeenCalledWith(
        { port },
        expect.any(Function),
      );
    });

    test("should create a HTTPS server on request", async () => {
      const configMock = {
        https: {
          listen: givePort(),
          options: { cert: "cert", key: "key" },
        },
        cors: true,
        startupLogo: false,
        logger: { level: "warn" as const },
      };
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            output: z.object({}),
            handler: vi.fn(),
          }),
        },
      };

      const { servers } = await createServer(
        createConfig(configMock),
        routingMock,
      );
      expect(servers).toHaveLength(1);
      expect(servers[0]).toBeTruthy();
      expect(createHttpsServerSpy).toHaveBeenCalledWith(
        configMock.https.options,
        appMock,
      );
      expect(httpsListenSpy).toHaveBeenCalledTimes(1);
      expect(httpsListenSpy).toHaveBeenCalledWith(
        configMock.https.listen,
        expect.any(Function),
      );
    });

    test("should create both HTTP and HTTPS servers", async () => {
      const configMock = {
        http: { listen: givePort() },
        https: {
          listen: givePort(),
          options: { cert: "cert", key: "key" },
        },
        cors: true,
        startupLogo: false,
        logger: { level: "warn" as const },
      };
      const { servers } = await createServer(createConfig(configMock), {});
      expect(servers).toHaveLength(2);
      expect(servers[0]).toBeTruthy();
      expect(servers[1]).toBeTruthy();
    });

    test("should enable compression on request", async () => {
      const configMock = {
        http: { listen: givePort() },
        compression: true,
        cors: true,
        startupLogo: false,
        logger: { level: "warn" },
      } satisfies ServerConfig;
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            output: z.object({}),
            handler: vi.fn(),
          }),
        },
      };
      await createServer(configMock, routingMock);
      expect(appMock.use).toHaveBeenCalledTimes(3);
      expect(compressionMock).toHaveBeenCalledTimes(1);
      expect(compressionMock).toHaveBeenCalledWith(undefined);
    });

    test("should enable uploads on request", async () => {
      const configMock = {
        http: { listen: givePort() },
        upload: {
          limits: { fileSize: 1024 },
          limitError: new Error("Too heavy"),
          beforeUpload: vi.fn(),
        },
        cors: true,
        startupLogo: false,
        logger: { level: "warn" },
      } satisfies ServerConfig;
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            input: z.object({
              file: ez.upload(),
            }),
            output: z.object({}),
            handler: vi.fn(),
          }),
        },
      };
      await createServer(configMock, routingMock);
      expect(appMock.use).toHaveBeenCalledTimes(2);
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/test",
        expect.any(Function), // uploader with logger
        expect.any(Function), // createUploadFailureHandler()
        expect.any(Function), // endpoint
      );
    });

    test("should enable raw on request", async () => {
      const configMock = {
        http: { listen: givePort() },
        cors: true,
        startupLogo: false,
        logger: { level: "warn" },
      } satisfies ServerConfig;
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            input: ez.raw(),
            output: z.object({}),
            handler: vi.fn(),
          }),
        },
      };
      await createServer(configMock, routingMock);
      expect(appMock.use).toHaveBeenCalledTimes(2);
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/test",
        expressRawMock,
        moveRaw,
        expect.any(Function), // endpoint
      );
    });
  });

  describe("attachRouting()", () => {
    test("should attach routing to the custom express app", () => {
      const app = express();
      expect(appMock).toBeTruthy();
      const customLogger = new BuiltinLogger({ level: "silent" });
      const infoMethod = vi.spyOn(customLogger, "info");
      const configMock = {
        app,
        cors: true,
        startupLogo: false,
        errorHandler: {
          handler: vi.fn(),
        },
        logger: customLogger,
      };
      const routingMock = {
        v1: {
          test: new EndpointsFactory(defaultResultHandler).build({
            methods: ["get", "post"],
            input: z.object({
              n: z.number(),
            }),
            output: z.object({
              b: z.boolean(),
            }),
            handler: vi.fn(),
          }),
        },
      };
      const { logger, notFoundHandler } = attachRouting(
        configMock as unknown as AppConfig,
        routingMock,
      );
      expect(logger).toEqual(customLogger);
      expect(typeof notFoundHandler).toBe("function");
      expect(appMock.use).toHaveBeenCalledTimes(1); // createLoggingMiddleware
      expect(configMock.errorHandler.handler).toHaveBeenCalledTimes(0);
      expect(infoMethod).toHaveBeenCalledTimes(0);
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get.mock.calls[0][0]).toBe("/v1/test");
      expect(appMock.post).toHaveBeenCalledTimes(1);
      expect(appMock.post.mock.calls[0][0]).toBe("/v1/test");
      expect(appMock.options).toHaveBeenCalledTimes(1);
      expect(appMock.options.mock.calls[0][0]).toBe("/v1/test");
    });
  });
});
