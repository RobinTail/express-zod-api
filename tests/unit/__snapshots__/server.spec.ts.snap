// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server createNotFoundHandler() the handler should call ResultHandler with 404 error 1`] = `
[
  {
    "error": [NotFoundError: Can not POST /v1/test],
    "input": null,
    "logger": {
      "debug": [MockFunction],
      "error": [MockFunction],
      "info": [MockFunction],
      "warn": [MockFunction],
    },
    "output": null,
    "request": {
      "body": {
        "n": 453,
      },
      "header": [MockFunction],
      "method": "POST",
      "path": "/v1/test",
    },
    "response": {
      "end": [MockFunction],
      "json": [MockFunction],
      "set": [MockFunction],
      "status": [MockFunction],
    },
  },
]
`;
